### PR TITLE
Added multi-state design capability with examples

### PR DESCRIPTION
**Pull request by hdieckhaus**
Added capability for multi-state design with MPNN.

Key points:
- Inputs for MSD can be one/multiple PDBs with labeled chains (ok if duplicate chain labels exist).
- Outputs are one/multiple FASTA/CSV files, as usual. MSD also generates a directory "msd" and an intermediate PDB.

First step: run generate_json_multi_state.py json_multi_state.flags
- This works like generate_json.py, but adds a few arguments and json components (most importantly, --constraints) and combines the multiple PDBs into an intermediate file called msd/msd.pdb.

Second step: run run_protein_mpnn.py proteinmpnn.flags
- Note that in proteinmpnn.flags, the --pdb_dir is the "msd" directory made in the first step so that it uses msd.pdb.
- ProteinMPNN does inference on the whole MSD system using tied_positions and tied_betas to enact the constraints.
- After inference, the MSD system is split back into separate files and saved as usual.
- I have also added per-chain sequence recovery and scoring when MSD is enabled.

Examples of positive, negative, and mixed design cases are provided under examples/multi_state

### CODE REVIEW

File: run/run_protein_mpnn.py [hunk 1]
```diff
@@ -179,6 +213,38 @@ def run_protein_mpnn(args):
...
+
+                                        scores_per_chain, seq_recovery_per_chain = {}, {}
+                                        for ctr, idx in zip(chains_original, chain_idx):
+                                                # grab the indices for scoring each chain
+                                                score_idx = chain_encoding_all == idx + 1
+                                                chain_score = torch.sum(scores_per_res[score_idx], dim=-1) / torch.sum(mask_for_loss[score_idx], dim=-1)
+                                                chain_score = np.format_float_positional(np.float32(chain_score.cpu().data.numpy().mean()), unique=False, precision=4)
+                                                scores_per_chain[ctr] = chain_score
+                                                # use this for seq recovery too
+                                                chain_seq_rec = torch.sum(seq_recovery_per_res.unsqueeze(0)[score_idx], dim=-1) / torch.sum(mask_for_loss[score_idx], dim=-1)
```
Comment (nzrandol) [inline run/run_protein_mpnn.py:236]: Is this `unsqueeze` in `seq_recovery_per_res.unsqueeze(0)[score_idx]` necessary? Couldn't we just do `seq_recovery_per_res`?

File: examples/multi_state/positive_design/mutres_redesign/outs/4GYT_dimer.csv
Comment (nzrandol) [inline examples/multi_state/positive_design/mutres_redesign/outs/4GYT_dimer.csv:1]: Should this file and the other .csv files been created? I don't see the `--af2_formatted_output` flag.

File: run/run_protein_mpnn.py
Comment (nzrandol) [inline run/run_protein_mpnn.py:1]: We might want to add the MSD changes to the `run_protein_mpnn_func` function since this is the function that EvoPro uses (I think)

File: run/generate_json_multi_state.py [hunk 2]
```diff
@@ -0,0 +1,374 @@
...
+                                fromfile_prefix_chars='@')
+
+    parser.add_argument('--pdb_dir',
+                        type=str,
+                        help='Path to the directory containing PDB files.')
+    parser.add_argument('--constraints', 
+                        type=str,
+                        default='',
+                        help='comma-separated list of multi-state design constraints. '
+                        'E.g. PDB1:A10-15:1+PDB2:A10-15:0.5+PDB3:B5-10:-3,PDB1:A20-25:1+PDB3:B20-25:1')
```
Comment (nzrandol) [inline run/generate_json_multi_state.py:342]: This says a comma-separated list, but from reading the code I think you can have multiple constraints listed here if they're separated by a semicolon (?). If this is true, we should add some documentation about that

**Review by nzrandol — CHANGES_REQUESTED** [review]
Overall, the changes look great! Could you please address some of the comments I left?

File: run/run_protein_mpnn.py (hunk 1 above)
Comment (hdieckhaus) [reply run/run_protein_mpnn.py:236]: Yes, the `score_idx` and `mask_for_loss` tensors have shape `[1, seq_length]`, but `seq_recovery_per_res` has shape `[seq_length],` so we get an `IndexError` if we just use `seq_recovery_per_res[score_idx]`.

So we can either add a dimension with unsqueeze:
`seq_recovery_per_res.unsqueeze(0)[score_idx]`
Or we can drop that extra dimension from `score_idx`:
`seq_recovery_per_res[torch.flatten(score_idx)]`

File: examples/multi_state/positive_design/mutres_redesign/outs/4GYT_dimer.csv
Comment (hdieckhaus) [reply examples/multi_state/positive_design/mutres_redesign/outs/4GYT_dimer.csv:1]: You're right, this must be left over from debugging. I will remove these and any others I find.
